File: web/lib/codex.mjs
import { readdir, stat, open } from 'node:fs/promises';
import { join } from 'node:path';
import { homedir } from 'node:os';

export const AGENTS = { claude: 'Claude Code', codex: 'Codex' };

/// The executable and arguments that reopen one session by the id its own transcript records.
export function resumeArguments(agent, sessionId) {
  return agent === 'codex' ? ['resume', sessionId] : ['--resume', sessionId];
}

export function agentCommand(agent, claude) {
  return agent === 'codex' ? 'codex' : claude;
}

export function codexRoot(env = process.env) {
  return env.SIFT_CODEX_ROOT || join(homedir(), '.codex', 'sessions');
}

// Codex prepends its own preamble to the user's words. Left in, every session in a repository
// ends up with the same title.
const INJECTED = [
  '# AGENTS.md instructions', '<permissions instructions>', '<INSTRUCTIONS>',
  '<environment_context>', '<user_instructions>', '<recommended_plugins>',
];

/// A thread Codex spawned for itself, the equivalent of Claude Code's sidechains. On the
/// library this was built against, 64 of 76 rollouts were these.
export function isSubagent(meta) {
  return meta?.thread_source === 'subagent' ||
    (meta?.source !== undefined && meta?.parent_thread_id !== undefined);
}

function textOf(content) {
  if (typeof content === 'string') return content.trim();
  if (!Array.isArray(content)) return '';
  return content
    .filter((b) => b && ['input_text', 'output_text', 'text'].includes(b.type) && b.text)
    .map((b) => b.text)
    .join('\n')
    .trim();
}

/// Codex writes one JSONL per session under `~/.codex/sessions/YYYY/MM/DD/`, nested by date
/// rather than by project. Every line is {timestamp, type, payload}. Returns null when the
/// file is a thread Codex spawned for itself, or holds nothing anyone said.
export async function parseTranscript(filePath) {
  const handle = await open(filePath, 'r');
  // Every field the store binds has to exist. An undefined one makes node:sqlite throw,
  // and the caller's catch turns that into a session that silently never appears.
  const meta = {
    sessionId: null, filePath, cwd: null, gitBranch: null, entrypoint: null, version: null,
    title: null, firstMessage: null, fullText: '', startedAt: null, lastActivity: null,
    messageCount: 0, toolCallCount: 0, agent: 'codex',
  };
  const texts = [];
  let subagent = false;

  try {
    for await (const line of handle.readLines({ encoding: 'utf8' })) {
      if (!line.trim()) continue;
      let obj;
      try { obj = JSON.parse(line); } catch { continue; }
      if (!obj || typeof obj !== 'object') continue;

      const at = obj.timestamp ? Date.parse(obj.timestamp) : NaN;
      if (!Number.isNaN(at)) {
        meta.startedAt ??= at;
        meta.lastActivity = at;
      }
      const payload = obj.payload;
      if (!payload || typeof payload !== 'object') continue;

      if (obj.type === 'session_meta') {
        if (isSubagent(payload)) subagent = true;
        meta.sessionId = payload.id || payload.session_id || null;
        meta.cwd = payload.cwd || null;
        meta.entrypoint = payload.originator || null;
        meta.version = payload.cli_version || null;
        meta.gitBranch = payload.git?.branch || null;
        continue;
      }
      if (obj.type !== 'response_item') continue;
      if (payload.type === 'function_call') { meta.toolCallCount += 1; continue; }
      if (payload.type !== 'message') continue;

      // `developer` is the harness talking to the model about sandboxing and policy, not one
      // of the two sides of the conversation.
      const role = payload.role;
      if (role !== 'user' && role !== 'assistant') continue;
      const text = textOf(payload.content);
      if (!text) continue;
      if (role === 'user' && INJECTED.some((p) => text.startsWith(p))) continue;

      meta.messageCount += 1;
      texts.push(text);
      if (role === 'user' && meta.firstMessage === null) meta.firstMessage = text.slice(0, 400);
    }
  } finally {
    await handle.close();
  }

  if (subagent || meta.messageCount === 0) return null;
  meta.fullText = texts.join('\n');
  meta.title = meta.firstMessage ? meta.firstMessage.slice(0, 90) : null;
  meta.sessionId ??= filePath.split(/[\\/]/).pop().replace(/\.jsonl$/, '');
  return meta;
}

export async function listTranscripts(root) {
  const out = [];
  async function walk(dir) {
    let entries;
    try { entries = await readdir(dir, { withFileTypes: true }); } catch { return; }
    for (const entry of entries) {
      const path = join(dir, entry.name);
      if (entry.isDirectory()) { await walk(path); continue; }
      if (!entry.name.startsWith('rollout-') || !entry.name.endsWith('.jsonl')) continue;
      try {
        const s = await stat(path);
        out.push({ filePath: path, projectId: 'codex', size: s.size, mtime: Math.floor(s.mtimeMs) });
      } catch { /* it can vanish between the listing and the stat */ }
    }
  }
  await walk(root);
  return out;
}

/// The conversation, for the reader pane.
export async function loadTurns(filePath, maxTurns = 200) {
  const handle = await open(filePath, 'r');
  const turns = [];
  try {
    for await (const line of handle.readLines({ encoding: 'utf8' })) {
      if (turns.length >= maxTurns) break;
      if (!line.trim()) continue;
      let obj;
      try { obj = JSON.parse(line); } catch { continue; }
      if (obj?.type !== 'response_item' || obj.payload?.type !== 'message') continue;
      const role = obj.payload.role;
      if (role !== 'user' && role !== 'assistant') continue;
      const text = textOf(obj.payload.content);
      if (!text) continue;
      if (role === 'user' && INJECTED.some((p) => text.startsWith(p))) continue;
      turns.push({ role, text, at: obj.timestamp ? Date.parse(obj.timestamp) : null });
    }
  } finally {
    await handle.close();
  }
  return turns;
}

File: windows-winui/Sift.Core/Codex.cs
using System.IO;
using System.Text.Json;

namespace SiftWinUI.Core;

public enum Agent { ClaudeCode, Codex }

public static class Agents
{
    public static string Code(this Agent agent) => agent == Agent.Codex ? "codex" : "claude";

    public static Agent Parse(string? code) => code == "codex" ? Agent.Codex : Agent.ClaudeCode;

    public static string Label(this Agent agent) => agent == Agent.Codex ? "Codex" : "Claude Code";

    /// The executable and the arguments that reopen one session by the id its own transcript
    /// records. Both agents take one, so a result goes back to the tool that wrote it.
    public static string Command(this Agent agent) => agent == Agent.Codex ? "codex" : "claude";

    public static string[] ResumeArguments(this Agent agent, string sessionId) =>
        agent == Agent.Codex ? ["resume", sessionId] : ["--resume", sessionId];
}

/// Codex writes one JSONL per session under `%USERPROFILE%\.codex\sessions\YYYY\MM\DD\`,
/// nested by date rather than by project. Every line is {timestamp, type, payload}: the first
/// is session_meta and carries the working directory and git branch, and the turns arrive as
/// response_item payloads of type message.
public static class Codex
{
    /// Codex prepends its own preamble to the user's words. Left in, every session in a
    /// repository ends up with the same title.
    private static readonly string[] Injected =
    [
        "# AGENTS.md instructions", "<permissions instructions>", "<INSTRUCTIONS>",
        "<environment_context>", "<user_instructions>", "<recommended_plugins>",
    ];

    public static string Root =>
        AppPaths.Resolve("SIFT_CODEX_ROOT")
        ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                        ".codex", "sessions");

    /// A thread Codex spawned for itself, the equivalent of Claude Code's sidechains. On the
    /// library this was built against, 64 of 76 rollouts were these; indexing them buries the
    /// sessions someone actually had.
    public static bool IsSubagent(JsonElement meta) =>
        (meta.TryGetProperty("thread_source", out var source) &&
         source.ValueKind == JsonValueKind.String && source.GetString() == "subagent") ||
        (meta.TryGetProperty("source", out _) && meta.TryGetProperty("parent_thread_id", out _));

    /// Returns null when the file is a subagent thread or holds nothing anyone said.
    public static SessionRow? ParseTranscript(string filePath)
    {
        var row = new SessionRow { FilePath = filePath };
        var full = new System.Text.StringBuilder();
        var isSubagent = false;

        foreach (var line in File.ReadLines(filePath))
        {
            if (string.IsNullOrWhiteSpace(line)) continue;
            JsonElement obj;
            try { obj = JsonDocument.Parse(line).RootElement; }
            catch (JsonException) { continue; }
            if (obj.ValueKind != JsonValueKind.Object) continue;

            var at = Millis(Str(obj, "timestamp"));
            if (at is not null)
            {
                row.StartedAt ??= at;
                row.LastActivity = at;
            }

            if (!obj.TryGetProperty("payload", out var payload) ||
                payload.ValueKind != JsonValueKind.Object) continue;

            switch (Str(obj, "type"))
            {
                case "session_meta":
                    if (IsSubagent(payload)) isSubagent = true;
                    row.SessionId = Str(payload, "id") is { Length: > 0 } id
                        ? id : Str(payload, "session_id");
                    row.Cwd = Blank(Str(payload, "cwd"));
                    row.Entrypoint = Blank(Str(payload, "originator"));
                    if (payload.TryGetProperty("git", out var git) &&
                        git.ValueKind == JsonValueKind.Object)
                        row.GitBranch = Blank(Str(git, "branch"));
                    break;

                case "response_item":
                    var kind = Str(payload, "type");
                    if (kind == "function_call") { row.ToolCallCount++; break; }
                    if (kind != "message") break;

                    // `developer` is the harness talking to the model about sandboxing and
                    // policy, not one of the two sides of the conversation.
                    var role = Str(payload, "role");
                    if (role != "user" && role != "assistant") break;
                    var text = TextOf(payload);
                    if (text.Length == 0) break;
                    if (role == "user" && Injected.Any(text.StartsWith)) break;

                    row.MessageCount++;
                    full.AppendLine(text);
                    if (role == "user" && row.FirstMessage is null)
                        row.FirstMessage = text[..Math.Min(400, text.Length)];
                    break;
            }
        }

        if (isSubagent || row.MessageCount == 0) return null;
        row.FullText = full.ToString();
        row.Title = row.FirstMessage?[..Math.Min(90, row.FirstMessage.Length)];
        if (row.SessionId.Length == 0) row.SessionId = Path.GetFileNameWithoutExtension(filePath);
        row.ProjectId = row.Cwd ?? "codex";
        return row;
    }

    public static List<Scanner.Found> ListTranscripts(string root)
    {
        var found = new List<Scanner.Found>();
        if (!Directory.Exists(root)) return found;

        foreach (var file in Directory.EnumerateFiles(root, "rollout-*.jsonl",
                                                      SearchOption.AllDirectories))
        {
            var info = new FileInfo(file);
            found.Add(new Scanner.Found(file, "codex", info.Length,
                                        new DateTimeOffset(info.LastWriteTimeUtc).ToUnixTimeMilliseconds()));
        }
        return found;
    }

    /// The conversation, for the reader pane.
    public static List<Turn> LoadTurns(string filePath, int maxTurns = 200)
    {
        var turns = new List<Turn>();
        foreach (var line in File.ReadLines(filePath))
        {
            if (turns.Count >= maxTurns) break;
            if (string.IsNullOrWhiteSpace(line)) continue;
            JsonElement obj;
            try { obj = JsonDocument.Parse(line).RootElement; }
            catch (JsonException) { continue; }
            if (obj.ValueKind != JsonValueKind.Object) continue;
            if (Str(obj, "type") != "response_item") continue;
            if (!obj.TryGetProperty("payload", out var payload)) continue;
            if (Str(payload, "type") != "message") continue;

            var role = Str(payload, "role");
            if (role != "user" && role != "assistant") continue;
            var text = TextOf(payload);
            if (text.Length == 0) continue;
            if (role == "user" && Injected.Any(text.StartsWith)) continue;
            turns.Add(new Turn(role, text, Millis(Str(obj, "timestamp"))));
        }
        return turns;
    }

    private static string TextOf(JsonElement payload)
    {
        if (!payload.TryGetProperty("content", out var content)) return "";
        if (content.ValueKind == JsonValueKind.String) return content.GetString()?.Trim() ?? "";
        if (content.ValueKind != JsonValueKind.Array) return "";

        var parts = new List<string>();
        foreach (var block in content.EnumerateArray())
        {
            if (block.ValueKind != JsonValueKind.Object) continue;
            if (Str(block, "type") is not ("input_text" or "output_text" or "text")) continue;
            var text = Str(block, "text");
            if (text.Length > 0) parts.Add(text);
        }
        return string.Join("\n", parts).Trim();
    }

    private static string Str(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString() ?? "" : "";

    private static string? Blank(string value) => value.Length == 0 ? null : value;

    private static long? Millis(string timestamp) =>
        DateTimeOffset.TryParse(timestamp, System.Globalization.CultureInfo.InvariantCulture,
                                System.Globalization.DateTimeStyles.AdjustToUniversal, out var parsed)
            ? parsed.ToUnixTimeMilliseconds() : null;
}
